### ISSUE TITLE
Crash on Resume - nil Screen Pointer Ref in awful/ewmh.lua

### ISSUE THREAD

**Issue by ewhac** [awesomeWM/awesome#2221]
Output of `awesome --version`:

awesome v4.2 (Human after all)
 • Compiled against Lua 5.3.3 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.5
 • LGI version: 0.9.2

**How to reproduce the issue:**

 * Close laptop lid; wait for LED to show entry into suspend mode.
 * Open laptop lid.

**Actual result:**

awesome has crashed; printing the following error messages.  The timestamps indicate the crash occurred during resume, not suspend:

```
2018-03-15 13:14:13 W: naughty.notify: there is no screen available to display the following notification: title='Oops, an error happened!' text='/usr/share/awesome/lib/awful/ewmh.lua:249: attempt to index a nil value (field 'screen')'
2018-03-15 13:14:13 W: awesome: luaA_dofunction:77: error while running function!
stack traceback:
	/usr/share/awesome/lib/awful/ewmh.lua:249: in function 'awful.ewmh.geometry'
	[C]: in method 'emit_signal'
	/usr/share/awesome/lib/awful/ewmh.lua:60: in function </usr/share/awesome/lib/awful/ewmh.lua:51>
error: /usr/share/awesome/lib/awful/ewmh.lua:249: attempt to index a nil value (field 'screen')
2018-03-15 13:14:13 W: naughty.notify: there is no screen available to display the following notification: title='Oops, an error happened!' text='/usr/share/awesome/lib/awful/ewmh.lua:249: attempt to index a nil value (field 'screen')'
2018-03-15 13:14:13 W: awesome: luaA_dofunction:77: error while running function!
stack traceback:
	/usr/share/awesome/lib/awful/ewmh.lua:249: in function 'awful.ewmh.geometry'
	[C]: in method 'emit_signal'
	/usr/share/awesome/lib/awful/ewmh.lua:60: in function </usr/share/awesome/lib/awful/ewmh.lua:51>
error: /usr/share/awesome/lib/awful/ewmh.lua:249: attempt to index a nil value (field 'screen')
2018-03-15 13:14:13 W: naughty.notify: there is no screen available to display the following notification: title='Oops, an error happened!' text='/usr/share/awesome/lib/awful/ewmh.lua:249: attempt to index a nil value (field 'screen')'
2018-03-15 13:14:13 W: awesome: luaA_dofunction:77: error while running function!
stack traceback:
	/usr/share/awesome/lib/awful/ewmh.lua:249: in function 'awful.ewmh.geometry'
	[C]: in method 'emit_signal'
	/usr/share/awesome/lib/awful/ewmh.lua:60: in function </usr/share/awesome/lib/awful/ewmh.lua:51>
error: /usr/share/awesome/lib/awful/ewmh.lua:249: attempt to index a nil value (field 'screen')
2018-03-15 13:14:13 W: awesome: a_glib_poll:432: Last main loop iteration took 0.318089 seconds! Increasing limit for this warning to that value.
2018-03-15 13:14:13 W: naughty.notify: there is no screen available to display the following notification: title='Oops, an error happened!' text='/usr/share/awesome/lib/awful/ewmh.lua:249: attempt to index a nil value (field 'screen')'
2018-03-15 13:14:13 W: awesome: luaA_dofunction:77: error while running function!
stack traceback:
	/usr/share/awesome/lib/awful/ewmh.lua:249: in function 'awful.ewmh.geometry'
error: /usr/share/awesome/lib/awful/ewmh.lua:249: attempt to index a nil value (field 'screen')
2018-03-15 13:14:13 W: awesome: luaA_dofunction:77: error while running function!
stack traceback:
	/usr/share/awesome/lib/awful/ewmh.lua:249: in function 'awful.ewmh.geometry'
error: /usr/share/awesome/lib/awful/ewmh.lua:249: attempt to index a nil value (field 'screen')
2018-03-15 13:14:13 W: awesome: luaA_dofunction:77: error while running function!
stack traceback:
	/usr/share/awesome/lib/awful/ewmh.lua:249: in function 'awful.ewmh.geometry'
error: /usr/share/awesome/lib/awful/ewmh.lua:249: attempt to index a nil value (field 'screen')
2018-03-15 13:14:13 E: awesome: signal_fatal:449: signal 11, dumping backtrace
awesome(backtrace_get+0x5a) [0x55d0c26beb19]
awesome(+0x12419) [0x55d0c26aa419]
/lib/x86_64-linux-gnu/libc.so.6(+0x34f00) [0x7f17eb5faf00]
awesome(screen_area_in_screen+0x4) [0x55d0c26d0b3d]
awesome(+0x2e0e2) [0x55d0c26c60e2]
awesome(client_resize+0xf1) [0x55d0c26c8b45]
awesome(+0x31261) [0x55d0c26c9261]
/usr/lib/x86_64-linux-gnu/liblua5.3.so.0(+0xd77e) [0x7f17ebbde77e]
/usr/lib/x86_64-linux-gnu/liblua5.3.so.0(+0x1a565) [0x7f17ebbeb565]
/usr/lib/x86_64-linux-gnu/liblua5.3.so.0(+0xdb6f) [0x7f17ebbdeb6f]
/usr/lib/x86_64-linux-gnu/liblua5.3.so.0(+0xdbc1) [0x7f17ebbdebc1]
/usr/lib/x86_64-linux-gnu/liblua5.3.so.0(+0xcf92) [0x7f17ebbddf92]
/usr/lib/x86_64-linux-gnu/liblua5.3.so.0(+0xde4d) [0x7f17ebbdee4d]
/usr/lib/x86_64-linux-gnu/liblua5.3.so.0(lua_pcallk+0x81) [0x7f17ebbda5c1]
awesome(signal_object_emit+0x267) [0x55d0c26c18ac]
awesome(luaA_class_emit_signal+0xd) [0x55d0c26c03b4]
awesome(luaA_object_emit_signal+0x3a7) [0x55d0c26c1c9a]
awesome(event_handle+0x452) [0x55d0c26afd3a]
awesome(+0x1236f) [0x55d0c26aa36f]
/lib/x86_64-linux-gnu/libglib-2.0.so.0(+0x4b119) [0x7f17ee6fb119]
/lib/x86_64-linux-gnu/libglib-2.0.so.0(g_main_loop_run+0xc2) [0x7f17ee6fb4b2]
awesome(main+0x150e) [0x55d0c26abbb1]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7) [0x7f17eb5e7a87]
awesome(_start+0x2a) [0x55d0c26aa04a]
```

**Expected result:**

awesome should resume without crash or restart.

**Additional Info:**

The system is a Thinkpad X1 Carbon (gen.4).  X11 is running with the 'glamoregl' driver.  Prior to last night, I have been running 4.2 without incident.  Last night, however, I updated a bunch of system libraries, and these crashes started happening.

System info you'd get if I'd submitted this through Debian's `reportbug`:

```
Debian Release: buster/sid
  APT prefers unstable
  APT policy: (500, 'unstable'), (500, 'testing'), (500, 'stable')
Architecture: amd64 (x86_64)

Kernel: Linux 4.15.0-1-amd64 (SMP w/4 CPU cores)
Locale: LANG=en_US.utf8, LC_CTYPE=en_US.utf8 (charmap=UTF-8), LANGUAGE=en_US.utf8 (charmap=UTF-8)
Shell: /bin/sh linked to /bin/dash
Init: systemd (via /run/systemd/system)
LSM: AppArmor: enabled

Versions of packages awesome depends on:
ii  dbus-user-session [default-dbus-session-bus]  1.12.6-2
ii  dbus-x11 [dbus-session-bus]                   1.12.6-2
ii  gir1.2-freedesktop                            1.54.1-4
ii  gir1.2-pango-1.0                              1.40.14-1
ii  libc6                                         2.27-2
ii  libcairo-gobject2                             1.15.10-2
ii  libcairo2                                     1.15.10-2
ii  libdbus-1-3                                   1.12.6-2
ii  libgdk-pixbuf2.0-0                            2.36.11-1
ii  libglib2.0-0                                  2.54.3-2
ii  liblua5.3-0                                   5.3.3-1
ii  libstartup-notification0                      0.12-5
ii  libx11-6                                      2:1.6.4-3
ii  libxcb-cursor0                                0.1.1-4
ii  libxcb-icccm4                                 0.4.1-1+b1
ii  libxcb-keysyms1                               0.4.0-1+b2
ii  libxcb-randr0                                 1.13-1
ii  libxcb-render0                                1.13-1
ii  libxcb-shape0                                 1.13-1
ii  libxcb-util0                                  0.3.8-3+b2
ii  libxcb-xinerama0                              1.13-1
ii  libxcb-xkb1                                   1.13-1
ii  libxcb-xrm0                                   1.0-3
ii  libxcb-xtest0                                 1.13-1
ii  libxcb1                                       1.13-1
ii  libxdg-basedir1                               1.2.0-1
ii  libxkbcommon-x11-0                            0.8.0-1
ii  libxkbcommon0                                 0.8.0-1
ii  lua-lgi                                       0.9.2-2
ii  menu                                          2.1.47+b1

Versions of packages awesome recommends:
ii  awesome-extra      2017110501
ii  feh                2.23.2-1
ii  rlwrap             0.43-1
ii  x11-xserver-utils  7.7+7+b1

Versions of packages awesome suggests:
ii  awesome-doc  4.2-4
```

**Comment by Elv13** [comment]
ref #1344 #2185

tl;dr: `c.screen` and `mouse.screen` cannot be `nil`. There is thousand of places in the code that assume it isn't `nil`. My potential solution for that is described in those issue, but isn't implemented and wont be for a long time given my "stuff to do" backlog.

However I guess we could just prevent the last screens from being deleted or freeze the Lua context when there is no screen as a temporary solution to avoid the crashes. It isn't a long term solution, but at least people will stop reporting the issue when for some buggy reason the resume code claims all screens are disconnected.

**Comment by psychon** [comment]
@Elv13 But still, awesome definitely should not crash here.

So, yeah, the problem here seems to be that all screens are removed by the X11 server. When we handle the last screen going away, `screen_removed()` will try to find a new screen for the client, but `screen_getbycoord()` returns `NULL`, so the screen is assigned a `NULL` pointer as screen. This was never intended to happen, so it is not surprising that the C code assumes in lots of places that `c->screen` is not a `NULL` pointer.

For this specific crash, I *guess* that the following might fix it (`screen_area_in_screen` does not work for `NULL` pointers).
```diff
diff --git a/objects/client.c b/objects/client.c
index f267ec57..63a17319 100644
--- a/objects/client.c
+++ b/objects/client.c
@@ -1680,7 +1685,7 @@ client_resize_do(client_t *c, area_t geometry)
     lua_State *L = globalconf_get_lua_State();
 
     screen_t *new_screen = c->screen;
-    if(!screen_area_in_screen(new_screen, geometry))
+    if(new_screen == NULL || !screen_area_in_screen(new_screen, geometry))
         new_screen = screen_getbycoord(geometry.x, geometry.y);
 
     /* Also store geometry including border */
```
I wouldn't be surprised when there were lots of other ways in which segfaults can happen for clients without a screen.

@ewhac Is it possible for you to test this patch? Sadly, I have no idea for a workaround.

**Comment by Elv13** [comment]
@psychon But what about pausing the event loop and freezing Lua. It's not clean, but better than a crash. The worst thing that can happen is that it never wakes up?

**Comment by psychon** [comment]
Without the event loop running, how do we notice that new screens appeared?

**Comment by Elv13** [comment]
The "classical way" would to be wait on a socket in the main thread while a worker thread wait for the event. When the worker thread detects the new screen, it write to the socket, waking up the main thread. The main thread could also have a second event loop. Both options are not very pretty and not deleting the last screen is probably simpler for little drawbacks.

**Comment by psychon** [comment]
Any suggestions for "other window managers"?

i3 seems to create a screen (they call it output) for the root window if all RandR outputs disappear: https://github.com/i3/i3/blob/fe95ac56bbc4de66b3b1a7120af587d72b033262/src/randr.c#L836-L844

**Comment by ewhac** [comment]
I applied the above commit to a local copy of master and built it.  It seems to be working; awesome enters and leaves suspend without incident.

**Comment by psychon** [comment]
Thanks a lot for testing

**Comment by ewhac** [comment]
I just noted an issue with this:

 * Open windows on screens 1, 4, and 5.
 * Suspend machine.
 * Resume machine.

All windows are now collected on screen 1.  (But at least nothing has crashed.)

Sorry I didn't notice this earlier.

**Comment by psychon** [comment]
@ewhac That's not specific to "all screens disappear", but happens whenever some screen is removed. The code "tries some things" and as a last resort, it just forces everything to the first tag of the first screen with tags that it fails.

Something with `tag.connect_signal("removal-pending", function(t) end)` in your config could improve the situation here. Actually... I'm amazed that I do not remember having seen this anywhere yet.
The following is untested, but should be a good starting point for own experiments.
```lua
tag.connect_signal("removal-pending", function(t)
    for s in screen do
        if s ~= t.screen then
            local new_t = awful.tag.find_by_name(s, t.name)
            if new_t then
                for c in pairs(t:clients()) do
                    c:tags{ new_t }
                end
                return
            end
        end
    end
end)
```

**Comment by ewhac** [comment]
Will play around with this; thank you.  However, since the default screen can now disappear on suspend, this is now an issue for all laptop users, not just those connecting and disconnecting external displays.

**Comment by blueyed** [comment]
@ewhac 
I wonder why your screen appears during suspend after all.
Is docking involved?
I am not seeing a `request::tag` signal on the client class myself when suspending/resuming.
